src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.State.cs

```diff
@@ -0,0 +1,90 @@
+            public string TypeName { get; set; }
+            public string DocumentName { get; set; }
+            public string TargetFileNameCandidate { get; set; }
+            public bool IsDocumentNameAValidIdentifier { get; set; }
+
+            private State(TService service, SemanticDocument document)
+            {
+                this._service = service;
+                this.SemanticDocument = document;
+            }
+
+            internal static State Generate(TService service, SemanticDocument document, TextSpan textSpan, CancellationToken cancellationToken)
+            {
+                var state = new State(service, document);
+                if (!state.TryInitialize(textSpan, cancellationToken))
+                {
+                    return null;
+                }
+
+                return state;
+            }
+
+            private bool TryInitialize(
+                TextSpan textSpan,
+                CancellationToken cancellationToken)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return false;
+                }
+
+                var tree = this.SemanticDocument.SyntaxTree;
+                var root = this.SemanticDocument.Root;
+                var syntaxFacts = this.SemanticDocument.Project.LanguageServices.GetService<ISyntaxFactsService>();
+
+                var typeDeclaration = _service.GetNodeToAnalyze(root, textSpan) as TTypeDeclarationSyntax;
+                if (typeDeclaration == null)
+                {
+                    return false;
+                }
+
+                var typeSymbol = this.SemanticDocument.SemanticModel.GetDeclaredSymbol(typeDeclaration, cancellationToken) as INamedTypeSymbol;
+
+                // compiler declared types, anonymous types, types defined in metadata should be filtered out.
+                if (typeSymbol == null ||
+                    typeSymbol.Locations.Any(loc => loc.IsInMetadata) ||
+                    typeSymbol.IsAnonymousType ||
+                    typeSymbol.IsImplicitlyDeclared)
+                {
+                    return false;
+                }
+
+                TypeNode = typeDeclaration;
+                TypeName = typeSymbol.Name;
+                DocumentName = Path.GetFileNameWithoutExtension(this.SemanticDocument.Document.Name);
+                IsDocumentNameAValidIdentifier = syntaxFacts.IsValidIdentifier(DocumentName);
+
+                // TODO: Make this check better, it won't detect Outer.Inner.cs cases.
+                if (string.Equals(DocumentName, TypeName, StringComparison.CurrentCulture))
+                {
+                    // if type name matches document name in a case sensitive manner, we have nothing more to do.
+                    return false;
+                }
+
+                TargetFileNameCandidate = Path.Combine(typeSymbol.Name + Path.GetExtension(this.SemanticDocument.Document.Name));
```

Comment:
added a comment. will add tests after the initial merge. i'm going to spend some dedicated time to testing. too many things to handle right now :)